src/reverse_iterator.hpp

```diff
@@ -0,0 +1,39 @@
+// Taken from https://gist.github.com/arvidsson/7231973
+
+#ifndef BITCOIN_REVERSE_ITERATOR_HPP
+#define BITCOIN_REVERSE_ITERATOR_HPP
+
+/**
+ * Template used for reverse iteration in C++11 range-based for loops.
+ * 
+ *   std::vector<int> v = {1, 2, 3, 4, 5};
+ *   for (auto x : reverse_iterate(v))
+ *       std::cout << x << " ";
+ */
+
+template <typename T>
+class reverse_range
+{
+    const T &x;
+    
+public:
+    reverse_range(const T &x) : x(x) {}
+    
+    auto begin() const -> decltype(this->x.rbegin())
+    {
```

Comment:
Yes, if I remove it I get ```‘begin’ function uses ‘auto’ type specifier without trailing return type```.
If I replace the auto with T::reverse_iterator I get ```need ‘typename’ before ‘T::reverse_iterator’ because ‘T’ is a dependent scope```.